### PR TITLE
azure_storage_table_entity: support propoerty data types

### LINKED ISSUE #5286: Unable to create azure table entity with a boolean field
_This issue was originally opened by @pedromfcampos as hashicorp/terraform#23757. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    

### Terraform Version

```
Terraform v0.12.18
```

### Terraform Configuration Files

I am trying to create a new azure table entity, following the [official documentation](https://www.terraform.io/docs/providers/azurerm/r/storage_table_entity.html), correctly setting resource group, storage account and storage table. I am also using a tfvars file with the following :

```hcl

tableEntitys= [
  {
    partition_key = "FirstPK"
    row_key       = "FirstRowKey"
    entity = {
      Value = false
    }
  },
  {
    partition_key = "SecondPK"
    row_key       = "SecondRowKey"
    entity = {
      Value = false
    }
  }
]

```

### Expected Behavior

The entities should have been created with the field Value with the Boolean type.

### Actual Behavior

Field Value has String type.

### Steps to Reproduce

terraform init and terraform apply.

### Workarounds

I've also tried this workaround:

```
tableEntitys= [
  {
    partition_key = "FirstPK"
    row_key       = "FirstRowKey"
    entity = {
      "Value@odata.type" = "Edm.Boolean"
      Value = false
    }
  },
  {
    partition_key = "SecondPK"
    row_key       = "SecondRowKey"
    entity = {
      "Value@odata.type" = "Edm.Boolean"
      Value = false
    }
  }
]
```

, in which i was successfull to create a boolean field in Azure, **HOWEVER**, terraform throws exception, making this workaround unacceptable:

```
Error: Error setting `entity` for Entity (Partition Key "    partition_key = "FirstPK"
" / Row Key "FirstRowKey") (Table "table" / Storage Account "storageaccount" / Resource Group "resourceGrouptest"): entity.Value: '' expected type 'string', got unconvertible type 'bool'

```

### PR DESCRIPTION
**Pull request by jblunck**
This change makes flattenEntity parse OData Data Types and persist them
in the Terraform state.

Fixes #5286 #12337

### CODE REVIEW

File: internal/services/storage/storage_table_entity_resource_test.go [hunk 1]
```diff
@@ -155,6 +177,26 @@ resource "azurerm_storage_table_entity" "test" {
...
+
+resource "azurerm_storage_table_entity" "test" {
+  storage_account_name = azurerm_storage_account.test.name
+  table_name           = azurerm_storage_table.test.name
+
+  partition_key = "test_partition%d"
+  row_key       = "test_row%d"
+  entity = {
+    Foo = 123
+	"Foo@odata.type" = "Edm.Int32"
```
Comment (katbyte) [inline internal/services/storage/storage_table_entity_resource_test.go:193]:
```suggestion
    "Foo@odata.type" = "Edm.Int32"
```

**Comment by jblunck** [comment]
@katbyte Is there anything I can do to move this PR forward?

**Comment by katbyte** [comment]
@jblunck - often just merging master in fixes such things

**Comment by jblunck** [comment]
> @jblunck - often just merging master in fixes such things

@katbyte Rebased to latest main branch.

**Comment by jblunck** [comment]
@katbyte Is there anything I can do to move this PR forward?

**Comment by Amier3** [comment]
Hey @jblunck 

Apologies for the late reply, volume has picked up a ton since this was made and this fell through the cracks. The acceptance tests are looking pretty good so i'm going to go ahead and merge this.

Thank you for the first time contribution and welcome to the terraform community! Hope to see more PRs from you in the future 👍 😄

**Pull request merged**